### ISSUE TITLE
`-Wundefined-func-template` warns that no definition is available for a pure virtual function

### ISSUE THREAD

**Issue by ahatanak** [llvm/llvm-project#74016]
The warning seems incorrect as `B::bar` is pure virtual and its definition isn't needed.
```
$ cat test.cpp
```
```cpp
template <typename T> class B {
public:
  constexpr void foo(const T &) { bar(1); }
  virtual constexpr void bar(unsigned int) = 0;
};

template <typename T> class D : public B<T> {
public:
  constexpr void bar(unsigned int) override {}
};

void test() {
  auto t = D<int>();
  t.foo(0);
}
```
```
$ clang++ -std=c++20 test.cpp -Wundefined-func-template -c
test.cpp:4:35: warning: instantiation of function 'B<int>::bar' required here, but no definition is available [-Wundefined-func-template]
  constexpr void foo(const T &) { bar(1); }
                                  ^
test.cpp:5:26: note: forward declaration of template entity is here
  virtual constexpr void bar(unsigned int) = 0;
                         ^
test.cpp:4:35: note: add an explicit instantiation declaration to suppress this warning if 'B<int>::bar' is explicitly instantiated in another translation unit
  constexpr void foo(const T &) { bar(1); }
                                  ^
1 warning generated.
```

**Comment by charmitro** [comment]
I'd like to pick this up. I'm sharing my thoughts so we can discuss them.

**Where does this happen?**

The warning is coming from https://github.com/llvm/llvm-project/blob/bc51a36fd67c9ef08259612a1a04db6be51a8b76/clang/lib/Sema/SemaTemplateInstantiateDecl.cpp#L4954-L4958

We do not check if the function is a pure virtual function via `FunctionDecl::isVirtualAsWritten()`.

IMO we should just add another condition in the `if` statement checking for this, such as:

```c++
      if (AtEndOfTU && !getDiagnostics().hasErrorOccurred() &&
          !getSourceManager().isInSystemHeader(PatternDecl->getBeginLoc()) &&
          !Function->isVirtualAsWritten()) { ... }
```

Finally we should add a test-case for this under [clang/test/SemaTemplate/](https://github.com/llvm/llvm-project/tree/main/clang/test/SemaTemplate) called `instantiate-virtual-function.cpp` (I'm not good at naming).

if all these sound good, I'll start implementing it.

**Comment by dwblaikie** [comment]
> if all these sound good, I'll start implementing it.

Sounds about right/plausible to me, thanks!

**Comment by ahatanaka** [comment]
We still want to emit the warning if `B::bar(1)` was called instead of `bar(1)` in `B::foo`, right?

The patch silences the warning.

**Comment by charmitro** [comment]
> We still want to emit the warning if `B::bar(1)` was called instead of `bar(1)` in `B::foo`, right?
> 
> The patch silences the warning.

Sorry  I didn't get this. Are we talking about the following?
```cpp
  constexpr void foo(const T &) { B::bar(1); }
```

instead of:
```cpp
  constexpr void foo(const T &) { bar(1); }
```

**Comment by ahatanak** [comment]
Yes, that's what I meant.

**Comment by charmitro** [comment]
> Yes, that's what I meant.

This is what we get with the patch:

```
$ ./bin/clang++ main.cc -Wundefined-func-template 
main.cc:10:26: warning: inline function 'B<int>::bar' is not defined [-Wundefined-inline]
   10 |   virtual constexpr void bar(unsigned int) = 0;
      |                          ^
main.cc:9:38: note: used here
    9 |   constexpr void foo(const T &) { B::bar(1); }
      |                                      ^
1 warning generated.
/usr/bin/ld: /tmp/main-c89dda.o: in function `B<int>::foo(int const&)':
main.cc:(.text._ZN1BIiE3fooERKi[_ZN1BIiE3fooERKi]+0x1a): undefined reference to `B<int>::bar(unsigned int)'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)

```

**Comment by charmitro** [comment]
> I meant the patch was silencing the `-Wundefined-func-template` warning.
> 
> Note that ToT clang emits the warning for the following code, but with the patch, it doesn't emit any warnings:
> 
> ```
> template <typename T> class B {
> public:
>   const void foo(const T &) { B::bar(1); }
>   virtual const void bar(unsigned int) = 0;
> };
> 
> template <typename T> class D : public B<T> {
> public:
>   const void bar(unsigned int) override {}
> };
> 
> void test() {
>   auto t = D<int>();
>   t.foo(0);
> }
> ```

Thanks for pinging me about this @ahatanak.

I revisited the patch based on your recommendation (thanks) and added a test case for it. 
Please check!
